crates/rust-analyzer/src/to_proto.rs

```diff
@@ -619,10 +619,41 @@ fn main() <fold>{
     }
 }
 
-pub(crate) fn code_action(world: &WorldSnapshot, assist: Assist) -> Result<lsp_ext::CodeAction> {
+pub(crate) fn unresolved_code_action(
+    world: &WorldSnapshot,
+    assist: Assist,
+) -> Result<lsp_ext::CodeAction> {
     let res = lsp_ext::CodeAction {
         title: assist.label,
-        group: if world.config.client_caps.code_action_group { assist.group_label } else { None },
+        id: Some(assist.id.0.to_owned()),
+        group: assist.group.and_then(|it| {
+            if world.config.client_caps.code_action_group {
+                None
+            } else {
+                Some(it.0)
+            }
+        }),
+        kind: Some(String::new()),
+        edit: None,
+        command: None,
+    };
+    Ok(res)
+}
+
+pub(crate) fn resolved_code_action(
+    world: &WorldSnapshot,
+    assist: ResolvedAssist,
+) -> Result<lsp_ext::CodeAction> {
+    let res = lsp_ext::CodeAction {
```

Comment:
Can we call `unresolved_code_action` and then clear the id and set the edit? To reuse code between the two functions.